src/masternodes/order.h

```diff
@@ -0,0 +1,222 @@
+    uint32_t closeHeight;
+
+    CFulfillOrderImplemetation()
+        : CFulfillOrder()
+        , creationTx()
+        , closeTx()
+        , creationHeight(-1)
+        , closeHeight(-1)
+    {}
+    ~CFulfillOrderImplemetation() override = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITEAS(CFulfillOrder, *this);
+        READWRITE(creationTx);
+        READWRITE(closeTx);
+        READWRITE(creationHeight);
+        READWRITE(closeHeight);
+    }
+};
+
+class CCloseOrder
+{
+public:
+    //! basic properties
+    uint256 orderTx;
+
+    CCloseOrder()
+        : orderTx(uint256())
+    {}
+    virtual ~CCloseOrder() = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(orderTx);
+    }
+};
+
+class CCloseOrderImplemetation : public CCloseOrder
+{
+public:
+    //! tx related properties
+    uint256 creationTx;
+    uint32_t creationHeight; 
+
+    CCloseOrderImplemetation()
+        : CCloseOrder()
+        , creationTx()
+        , creationHeight(-1)
+    {}
+    ~CCloseOrderImplemetation() override = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITEAS(CCloseOrder, *this);
+        READWRITE(creationTx);
+        READWRITE(creationHeight);
+    }
+};
+
+class COrderView : public virtual CStorageView {
+public:
+    typedef std::pair<DCT_ID,DCT_ID> TokenPair;
+    typedef std::pair<TokenPair,uint256> TokenPairKey;
+
+    using COrderImpl = COrderImplemetation;
+    using CFulfillOrderImpl = CFulfillOrderImplemetation;
+    using CCloseOrderImpl = CCloseOrderImplemetation;
+
+    std::unique_ptr<COrderImpl> GetOrderByCreationTx(const uint256 & txid) const;
+    ResVal<uint256> CreateOrder(const COrderImpl& order);
+    ResVal<uint256> CloseOrderTx(const uint256& txid);
+    void ForEachOrder(std::function<bool (TokenPairKey const &, CLazySerialize<COrderImpl>)> callback, TokenPair const & pair=TokenPair());    
+    
+    template<typename By, typename KeyType, typename ValueType>
+    bool ForEachOrder(std::function<bool(KeyType const &, CLazySerialize<ValueType>)> callback, KeyType const & start = KeyType()) const {
+        auto& self = const_cast<COrderView&>(*this);
+        auto it = self.DB().NewIterator();        
+        auto key = std::make_pair(By::prefix, start);
+        for(it->Seek(DbTypeToBytes(key)); it->Valid() && BytesToDbType(it->Key(), key) && key.first == By::prefix; it->Next())
+        {
+            boost::this_thread::interruption_point();
+            if ((start!=KeyType() && key.second.first!=start.first) || !callback(key.second, CLazySerialize<COrderImpl>(*it)))
+                break;
+        }
+        return true;
```

Comment:
It should go away it has a function that do exactly that.
```
struct TokenPairKey {
DCT_ID token1;
DCT_ID token2;
uint256 txid;
+ serialize methods
};
Partial key is TokenPairKey key = { {}, {}, txid };
```

Reply:
Which function is that? I didn't found it.